src/directory/file_watcher.rs

```diff
@@ -0,0 +1,136 @@
+use crate::directory::{WatchCallback, WatchCallbackList, WatchHandle};
+use crc32fast::Hasher;
+use std::fs;
+use std::io;
+use std::io::BufRead;
+use std::path::PathBuf;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+
+pub const DEFAULT_POLLING_INTERVAL: Duration = Duration::from_millis(500);
+
+// Watches a file and executes registered callbacks when the file is modified.
+pub struct FileWatcher {
+    path: Arc<PathBuf>,
+    callbacks: Arc<WatchCallbackList>,
+    state: Arc<AtomicUsize>, // 0: new, 1: runnable, 2: terminated
+}
+
+impl FileWatcher {
+    pub fn new(path: &PathBuf) -> FileWatcher {
+        FileWatcher {
+            path: Arc::new(path.clone()),
+            callbacks: Default::default(),
+            state: Default::default(),
+        }
+    }
+
+    pub fn spawn(&self) {
+        if self.state.compare_and_swap(0, 1, Ordering::Relaxed) > 0 {
+            return;
+        }
+
+        let path = self.path.clone();
+        let callbacks = self.callbacks.clone();
+        let state = self.state.clone();
+
+        thread::Builder::new()
+            .name("thread-tantivy-meta-file-watcher".to_string())
+            .spawn(move || {
+                let mut current_checksum = None;
+
+                while state.load(Ordering::Relaxed) == 1 {
+                    if let Ok(checksum) = FileWatcher::compute_checksum(&path) {
+                        if current_checksum.is_none() || current_checksum.unwrap() != checksum {
```

Comment:
i prefer `if let Some(current_checksum)= current_checksum_opt` to make the it visible that this code cannot panic.